portal/csrf.py

```diff
@@ -0,0 +1,57 @@
+from flask import abort, current_app, Blueprint, request
+from flask_wtf.csrf import CSRFProtect
+
+from .models.user import current_user
+
+csrf = CSRFProtect()
+
+
+csrf_blueprint = Blueprint('csrf_blueprint', __name__)
+
+
+@csrf_blueprint.before_app_request
+def csrf_protect():
+    """ Protect views from Cross Site Resource Forgery (CSRF)
+
+    All oauth endpoints must include the `@csrf.exempt` decorator in order
+    to avoid requiring csrf tokens for legitamite OAuth client use.
+
+    Given that many of these same endpoints are also used by the UI and other
+    user agent access (READ: vulnerable to CSRF attacks), this function is
+    invoked before every request (regardless of blueprint) to determine if csrf
+    protection should be included.
+
+    """
+    # Only protect the configured verbs
+    if request.method not in current_app.config['WTF_CSRF_METHODS']:
+        return
+
+    # Don't get in the way of the initial oauth dance.
+    if request.path.startswith('/oauth/'):
+        return
+
+    # Backdoor for testing
+    if current_app.config.get('TESTING') == True:
+        return
+
+    # Look for legit OAuth requests, and exclude these from csrf protection
+    if request.headers and request.headers.get('Authorization'):
```

Comment:
i sure would think it would always be there - just being careful given the frequent use of this method in the call path. fair point - i'll eliminate the redundant check.